### PR TITLE
[Refactor][Ops] align W0 ops with manifest parity check

### LINKED ISSUE #1376: [REFACTOR][OPS] add parity overrides / opt-outs for newly-aligned ops (W3-01)
## Description

`scripts/validate_manifest.py` runs a parity check between manifest spec and op-layer code (output-shape inference, dtype contract). For ops at `status: spec-only`, parity is skipped. After the W2 flip issues merged, ~78 newly-aligned ops in elementwise / reduction / normalization families became `implemented`. The validator's parity layer expects each op to expose `_infer_output_shapes` / `_validate_dtypes` overrides whose result matches the manifest's `shape_rules` / `dtype_combos` declaration.

**Updated 2026-05-09**: the previous brief referenced `parity_opt_out` / `parity_opt_out_reason` manifest fields that were retired in PR #1292 and are now schema-rejected at `scripts/validate_manifest.py:399-407`. The remaining "override on every op" arm spanned 83 ops × 2 methods, beyond a single PR. This brief is now scoped down to the **8 W0-01 ops** as a first per-family slice; further families should be filed as separate follow-ups.

The 8 W0 ops:
`ClampFwdOp`, `ClampScalarFwdOp`, `ClampMinFwdOp`, `ClampMaxFwdOp`, `MaskedFillFwdOp`, `MaskedFillScalarFwdOp`, `WhereFwdOp`, `RMSNormFwdOp`.

### Related Files

- `tileops/ops/op_base.py` — defines `_infer_output_shapes` / `_validate_dtypes` hooks
- `tileops/ops/elementwise.py` (or `tileops/ops/elementwise/*.py` after split) — Clamp / MaskedFill / Where Op classes
- `tileops/ops/norm/rms_norm.py` — `RMSNormFwdOp`
- `scripts/validate_manifest.py` — parity check entry point

## Goal

`python scripts/validate_manifest.py` emits zero parity warnings for the 8 listed ops. Each op either has both override methods aligned with manifest, or the op is demoted to `status: spec-only` with a one-line reason in the PR description (since `parity_opt_out` is no longer a valid manifest field).

## Plan

1. Run the validator's strict-parity mode and enumerate parity warnings for the 8 W0 ops only.
2. Classify each warning: **mechanical** (override mirrors `shape_rules`) or **indeterminate** (output truly cannot be inferred from inputs alone).
3. For mechanical: add `_infer_output_shapes` and/or `_validate_dtypes` on the op class.
4. For indeterminate: demote the op's manifest entry to `status: spec-only` (status flip is allowed under the carve-out) and document the demotion in the PR description.
5. Re-run validator; assert zero parity warnings on the 8 ops.

## Constraints

- MUST NOT modify manifest fields except `status` (the carve-out).
- MUST NOT change op runtime behavior. Overrides are pure introspection; existing tests must pass unchanged.
- MUST NOT use `parity_opt_out` / `parity_opt_out_reason` (retired schema; will fail validation).
- Scope strictly the 8 W0 ops. Other family overrides are out of scope.
- Single PR.

## Acceptance Criteria

- AC-1: Modified files pass unit tests (`pytest tests/ops/`).
- AC-2: `python scripts/validate_manifest.py` emits zero parity warnings for the 8 W0 ops.
- AC-3: `tests/test_validate_manifest.py` passes.
- AC-4: Each op is either (a) covered by both `_infer_output_shapes` and `_validate_dtypes` overrides aligned with manifest, OR (b) demoted to `status: spec-only` with a one-line reason in the PR description.
- AC-5: All 8 W0 ops are explicitly addressed (one row in the PR description per op showing the chosen approach).

### PR DESCRIPTION
**Pull request by lcy-seso**
Closes #1376

## Summary

- Align 8 W0 ops with the manifest parity check (default `validate_manifest.py` mode reports zero `[shape]/[dtype]` warnings for all 8).
- `WhereFwdOp`: add mechanical `_infer_output_shapes` and `_validate_dtypes` overrides matching the manifest broadcast/dtype contract.
- Demote `MaskedFillFwdOp`, `MaskedFillScalarFwdOp`, and `RMSNormFwdOp` to `status: spec-only` with documented per-op reasons in their manifest entries.
- Clamp family (`ClampFwdOp`, `ClampScalarFwdOp`, `ClampMinFwdOp`, `ClampMaxFwdOp`) was already aligned — no code change needed.

## Per-op approach

| Op | Approach | Notes |
| --- | --- | --- |
| `ClampFwdOp` | already aligned | overrides match manifest; no change |
| `ClampScalarFwdOp` | already aligned | overrides match manifest; no change |
| `ClampMinFwdOp` | already aligned | overrides match manifest; no change |
| `ClampMaxFwdOp` | already aligned | overrides match manifest; no change |
| `WhereFwdOp` | mechanical override added | `_infer_output_shapes` + `_validate_dtypes` aligned with manifest broadcast/dtype contract |
| `MaskedFillFwdOp` | demoted to `spec-only` | implementation deviates from spec; flip in follow-up impl PR |
| `MaskedFillScalarFwdOp` | demoted to `spec-only` | implementation deviates from spec; flip in follow-up impl PR |
| `RMSNormFwdOp` | demoted to `spec-only` | implementation deviates from spec; flip in follow-up impl PR |

## Test plan

- [x] Modified files pass unit tests (pytest tests/ops/) — 318 passed across the changed-op test files (`test_elementwise_compile.py`, `test_special_elementwise.py`, `test_special_elementwise_conformance.py`, `test_elementwise_independent_fp8.py`, `test_elementwise_unary_activation_alignment.py`, `test_rms_norm.py`, `test_normalization_alignment.py`).
- [x] `python scripts/validate_manifest.py` emits zero parity warnings for the 8 W0 ops (default mode).
- [x] `tests/test_validate_manifest.py` passes — 218 passed.
- [x] Each op is either covered by both overrides aligned with manifest, or demoted to `status: spec-only` with reason in this description.
- [x] All 8 W0 ops are explicitly addressed in the per-op approach table above.

### CODE REVIEW

**Review by Ibuki-wind — APPROVED** [review]
Clean — no issues.

File: tileops/ops/elementwise.py [hunk 1]
```diff
@@ -2398,6 +2398,42 @@ def __init__(
...
+        input_shape: tuple,
+        other_shape: tuple,
+    ) -> Dict[str, tuple]:
+        out = tuple(
+            torch.broadcast_shapes(
+                tuple(condition_shape),
+                tuple(input_shape),
+                tuple(other_shape),
+            )
+        )
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tileops/ops/elementwise.py]: Not assessed on latest HEAD

File: tileops/ops/elementwise.py [hunk 2]
```diff
@@ -2398,6 +2398,42 @@ def __init__(
...
+        return {"output": out}
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tileops/ops/elementwise.py]: Not assessed on latest HEAD

File: tileops/ops/elementwise.py [hunk 3]
```diff
@@ -2398,6 +2398,42 @@ def __init__(
...
+    ) -> None:
+        if condition.dtype != torch.bool:
+            raise ValueError(
+                f"Expected condition.dtype torch.bool, got {condition.dtype}"
+            )
+        if input.dtype not in self._SUPPORTED_DTYPES:
+            names = ", ".join(str(dt) for dt in self._SUPPORTED_DTYPES)
+            raise ValueError(
+                f"Expected input.dtype in [{names}], got {input.dtype}"
+            )
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tileops/ops/elementwise.py]: Not assessed on latest HEAD

File: tileops/manifest/normalization.yaml [hunk 4]
```diff
@@ -7,7 +7,7 @@
 RMSNormFwdOp:
   ref_api: "torch.nn.functional.rms_norm"
   family: normalization
-  status: implemented
+  status: spec-only
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tileops/manifest/normalization.yaml:11]: Fixed in 5765739 — added inline `# spec-only: <reason>` comment above each demoted op's status line.

File: tileops/manifest/elementwise_binary.yaml [hunk 5]
```diff
@@ -60,7 +60,7 @@ MaskedFillFwdOp:
   # value form lands as MaskedFillScalarFwdOp below per the
   # "No Optional[Tensor]" manifest rule for variant splits.
   family: elementwise
-  status: implemented
+  status: spec-only
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tileops/manifest/elementwise_binary.yaml:65]: Fixed in 5765739 — same as sibling MaskedFill thread.

File: tileops/manifest/elementwise_binary.yaml [hunk 6]
```diff
@@ -97,7 +97,7 @@ MaskedFillScalarFwdOp:
   # Per the "No Optional[Tensor]" manifest rule, this is split from the
   # 0-dim-Tensor-value primary entry above.
   family: elementwise
-  status: implemented
+  status: spec-only
   variant_of: MaskedFillFwdOp
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tileops/manifest/elementwise_binary.yaml:103]: Fixed in 5765739 — RMSNormFwdOp now has `# spec-only: shape_rules reference runtime-bound normalized_shape param without manifest default` above its status line.
